### PR TITLE
Add check for Application subresource under a Source during Availability check

### PR DESCRIPTION
**Pull request by lindgrenj6**
Should solve https://projects.engineering.redhat.com/browse/TPINVTRY-988

After discussions about whether an Endpoint is necessary on a Source or not we came ot the conclusion that we should really treat the Source that an Application lives under as a different "type" of source, i.e. some sources have endpoints, others have applications. This will check if there is an application present if there is no endpoint. 

\# TODO:
- [x] specs

### CODE REVIEW

File: lib/topological_inventory/providers/common/operations/sources_api_client.rb [hunk 1]
```diff
@@ -25,6 +25,11 @@ def fetch_default_endpoint(source_id)
             endpoints.find(&:default)
           end
 
+          def fetch_application(source_id)
+            applications = api.list_source_applications(source_id)&.data || []
+            applications.first
```
Comment (slemrmartin) [inline lib/topological_inventory/providers/common/operations/sources_api_client.rb:30]: as we are checking first endpoint (as default), there is no such concept in applications. 
shouldn't we check all of them?

File: lib/topological_inventory/providers/common/operations/source.rb [hunk 2]
```diff
@@ -57,32 +57,57 @@ def params_missing?
...
+
+        def endpoint_connection_check
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+
+          # call down into the operations pod implementation of `Source#connection_check`
           connection_check
         end
 
+        def application_connection_check
+          case application.availability
```
Comment (slemrmartin) [inline lib/topological_inventory/providers/common/operations/source.rb]: who is setting this value?
Is it by https://github.com/RedHatInsights/sources-api/blob/master/app/controllers/api/v1/sources_controller.rb#L56 call?

Comment (slemrmartin) [reply lib/topological_inventory/providers/common/operations/source.rb]: btw I don't see `application.availability` in Sources API

Comment (lindgrenj6) [reply lib/topological_inventory/providers/common/operations/source.rb]: Yeah I typo'd that, it's updated now!

Comment (lindgrenj6) [reply lib/topological_inventory/providers/common/operations/source.rb]: also I talked Doug in cost, they're updating it from their end every time an availability request comes in.

File: lib/topological_inventory/providers/common/operations/sources_api_client.rb (hunk 1 above)
Comment (lindgrenj6) [reply lib/topological_inventory/providers/common/operations/sources_api_client.rb:30]: Hmmm yeah good question, should we fail the whole check if one of them is unavailable? Or only fail if _all_ of them are unavailable?

Comment (slemrmartin) [reply lib/topological_inventory/providers/common/operations/sources_api_client.rb:30]: yeah, also good questions. This is the "Partially available" problem. @gmcculloug ?

Comment (syncrou) [reply lib/topological_inventory/providers/common/operations/sources_api_client.rb:30]: For help on the original thoughts. Here is the Sources Availability matrix put together by Richard: https://github.com/RedHatInsights/sources-ui/issues/198

**Review by slemrmartin — APPROVED** [review]
Regarding our meeting, this is good for Cost Management team's issue, we'll handle more difficult cases in following PR(s)

**Review by syncrou — APPROVED** [review]
LGTM 👍 

Will do a follow up on handling the 1 -> many objects

**Pull request merged**